### ISSUE TITLE
Ideas

### ISSUE THREAD

**Issue by klzgrad** [klzgrad/naiveproxy#503]
https://github.com/net4people/bbs/issues/136#issuecomment-1279687386

* per tunnel preamble: Perform a real or realistic http request as the tunnel preamble to confuse traffic analysis systems that handles the starting part of a traffic flow (i.e. most of them). Issues: this adds connection latency, but can be amortized with multiplexing.
* Tunnel rotation
* per stream cover traffic, depending on multiplexing activities
* A traffic collection tool
  * chrome://net-export with script
  * wireshark with script
  * Tool built in
* traffic flow visualizer
* packet length distribution visualizer

**Comment by triggered96** [comment]
> [net4people/bbs#136 (comment)](https://github.com/net4people/bbs/issues/136#issuecomment-1279687386)
> 
> * Per connection preamble: Perform a real or realistic http request as the tunnel preamble to confuse traffic analysis systems that handles the starting part of a traffic flow (i.e. most of them). Issues: this adds connection latency, but can be amortized with multiplexing. [Large scale blocking of TLS-based censorship circumvention tools in China net4people/bbs#129 (comment)](https://github.com/net4people/bbs/issues/129#issuecomment-1272590029)
> * Per stream cover traffic: Add background covering traffic if there is no other activities in the current multiplexed connection.
> * Connection rotation: Limit the lifetime of each connection, helps CGNAT, Iran.
> * Backup connection with preamble warmup to reduce latency.
> * Coalesces consecutive small writes into single larger messages when there are multiple pending writes (see https://pkg.go.dev/github.com/getlantern/lampshade), with small amount of corking
> * Implicit padding lengths generated from a seed: Saves padding layering overhead
> * Realistic traffic profile
> * A traffic collection tool
>   
>   * chrome://net-export with script
>   * Wireshark with script
>   * Tool built in stats
> * Traffic flow visualizer
> * Packet length distribution visualizer
> * Chromium net API [Stripping down browser forwarding with external Chrome stack v2fly/v2ray-core#2092](https://github.com/v2fly/v2ray-core/discussions/2092)

I agree with the author's approach, it's time to update. Today's naive is almost more and more likely to be detected by GFW.

**Comment by Lingxi-Li** [comment]
Thought the artificial traffic could be generated against the camouflage site. E.g., if the camo site is a file server, the traffic could be generated by browsing within the file server.

**Comment by univerioiln233** [comment]
blocking few mtr echo-request:
-A ufw-before-input -p icmp --icmp-type echo-request -j DROP

**Comment by alidxdydz** [comment]
Dear @klzgrad

I'm an end-user and don't have technical network knowledge, but I thought you may find this project helpful.

I realized you use similar literature regarding "characteristics of network traffic" as Dust [v1](https://raw.githubusercontent.com/blanu/Dust/master/historical/v1/README) and [v2](https://raw.githubusercontent.com/blanu/Dust/master/historical/haskell2014/Dust/README) which was [presented in DEF CON](https://youtu.be/u0iueUlY6dQ).
I thought Dust project was abandoned and its original author didn't maintain it. However, I found out that there is a [successor project](https://github.com/OperatorFoundation) maintained by the original author.

So far, they have released Replicant ([go](https://github.com/OperatorFoundation/Replicant-go), [swift](https://github.com/OperatorFoundation/ReplicantSwift)), Starbridge ([go](https://github.com/OperatorFoundation/Starbridge-go), [swift](https://github.com/OperatorFoundation/Starbridge)), Shadow ([go](https://github.com/OperatorFoundation/Shadow-go), [swift](https://github.com/OperatorFoundation/ShadowSwift)), and Optimizer ([go](https://github.com/OperatorFoundation/Optimizer-go)) as Shapeshifter [Dispatchers](https://github.com/OperatorFoundation/shapeshifter-dispatcher) and [Transports](https://github.com/OperatorFoundation/shapeshifter-transports).

Also, they have developed [Flower](https://github.com/OperatorFoundation/Flower) protocol with its server-side [Persona](https://github.com/OperatorFoundation/Persona) implementation and client-side [Moonbounce](https://github.com/OperatorFoundation/Moonbounce) implementation for macOS only, emphasizing on its built-in shapeshifting technology.

I don't know why there is no technical discussion about this project in [bbs](https://github.com/net4people/bbs/issues) and other censorship circumvention communities and whether it works in China and Russia, and Iran, but I hope these can be facilitate what you are trying to achieve.

**Comment by klzgrad** [comment]
I'm aware of the literature you referenced.

It's fine to not have discussion about this project. This is focused on usability and performance, not research novelty and research-wise it's actually quite naive (can trace the root to meek, but with less resource requirement). It was by accident that it became useful for some, and it does not need unwanted attention.

**Comment by klzgrad** [comment]
https://github.com/klzgrad/naiveproxy/releases/tag/v147.0.7727.49-3
